sycl/include/std/experimental/simd.hpp

```diff
@@ -0,0 +1,1740 @@
+template <class Abi> simd_div_t<shortv<Abi>> div(shortv<Abi> numer, shortv<Abi> denom);
+template <class Abi> simd_div_t<intv<Abi>> div(intv<Abi> numer, intv<Abi> denom);
+template <class Abi> simd_div_t<longv<Abi>> div(longv<Abi> numer, longv<Abi> denom);
+template <class Abi> simd_div_t<llongv<Abi>> div(llongv<Abi> numer, llongv<Abi> denom);
+
+// [simd.mask.class]
+template <class T, class Abi>
+class simd_mask {
+public:
+  using value_type = bool;
+  using reference = see below;
+  using simd_type = simd<T, Abi>;
+  using abi_type = Abi;
+  static constexpr size_t size() noexcept;
+  simd_mask() = default;
+
+  // broadcast constructor
+  explicit simd_mask(value_type) noexcept;
+
+  // implicit type conversion constructor
+  template <class U> simd_mask(const simd_mask<U, simd_abi::fixed_size<size()>>&) noexcept;
+
+  // load constructor
+  template <class Flags> simd_mask(const value_type* mem, Flags);
+
+  // loads [simd.mask.copy]
+  template <class Flags> void copy_from(const value_type* mem, Flags);
+  template <class Flags> void copy_to(value_type* mem, Flags) const;
+
+  // scalar access [simd.mask.subscr]
+  reference operator[](size_t);
+  value_type operator[](size_t) const;
+
+  // unary operators [simd.mask.unary]
+  simd_mask operator!() const noexcept;
+
+  // simd_mask binary operators [simd.mask.binary]
+  friend simd_mask operator&&(const simd_mask&, const simd_mask&) noexcept;
+  friend simd_mask operator||(const simd_mask&, const simd_mask&) noexcept;
+  friend simd_mask operator& (const simd_mask&, const simd_mask&) noexcept;
+  friend simd_mask operator| (const simd_mask&, const simd_mask&) noexcept;
+  friend simd_mask operator^ (const simd_mask&, const simd_mask&) noexcept;
+
+  // simd_mask compound assignment [simd.mask.cassign]
+  friend simd_mask& operator&=(simd_mask&, const simd_mask&) noexcept;
+  friend simd_mask& operator|=(simd_mask&, const simd_mask&) noexcept;
+  friend simd_mask& operator^=(simd_mask&, const simd_mask&) noexcept;
+
+  // simd_mask compares [simd.mask.comparison]
+  friend simd_mask operator==(const simd_mask&, const simd_mask&) noexcept;
+  friend simd_mask operator!=(const simd_mask&, const simd_mask&) noexcept;
+};
+
+} // parallelism_v2
+} // std::experimental
+
+*/
+
+// Added for SYCL/invoke_simd {
+#define _LIBCPP_STD_VER 17
+#define _LIBCPP_COMPILER_CLANG_BASED 1
+#define _LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER 1
+#define _LIBCPP_BEGIN_NAMESPACE_EXPERIMENTAL_SIMD namespace std::experimental {
+#define _LIBCPP_END_NAMESPACE_EXPERIMENTAL_SIMD }
+#define _LIBCPP_BEGIN_NAMESPACE_EXPERIMENTAL_SIMD_ABI namespace std::experimental::simd_abi {
+#define _LIBCPP_END_NAMESPACE_EXPERIMENTAL_SIMD_ABI }
+#define _LIBCPP_INLINE_VAR inline
+#define _LIBCPP_PUSH_MACROS
+#define _LIBCPP_POP_MACROS
+#define _LIBCPP_NATIVE_SIMD_WIDTH_IN_BYTES 512 // is not really used for now for sycl::ext::oneapi::experimental::invoke_simd
```

Comment:
This should eventually be device-specific, right?  Do we need this long-term, and if so do we have any plans for how to make it device-specific?

Reply:
Yes, I think so. This is used to define the `simd_abi::native`, which is different for different devices.
`-target XXX` clang option where XXX != spir64 should define `_LIBCPP_NATIVE_SIMD_WIDTH_IN_BYTES` according to XXX 
target capabilities. This is also true for device-specific SYCL AOT targets (e.g. `spir64_gen`).
But for device-neutral `spir64` there is no perfect choice. There will have to be some reasonable default, but I'd expect 
SYCL people won't use it. Instead, they would use `oneapi::simd_abi::native_fixed_size` with proper kernel versioning based 
on [[sycl::requires]].

I think in oneapi we could introduce another kind of native ABI templated by a SYCL aspect corresponding to a device architecture (which will soon appear as an extension, I believe).